### PR TITLE
Address accuracy issues introduced by fp16 packed arithmetic

### PR DESCRIPTION
**Pull request by giuseros**
Fixes https://github.com/ROCm/rocMLIR-internal/issues/1466

### CODE REVIEW

**Review by manupak — DISMISSED** [review]
LGTM on the basis that it is an accepted trade off to use packed f16 arithmetic!

File: mlir/lib/ExecutionEngine/conv-validation-wrappers.cpp [hunk 1]
```diff
@@ -484,6 +484,9 @@ void mcpuVerify(T *gpuResults, T *validationResults, long long dataSize,
 
     if (valNum == gpuNum) {
       hist_relDiff[0]++;
+    } else if (std::isinf(valNum)) {
```
Comment (manupak) [inline mlir/lib/ExecutionEngine/conv-validation-wrappers.cpp]: If the validation result == inf, then we accept any error ?

Comment (manupak) [reply mlir/lib/ExecutionEngine/conv-validation-wrappers.cpp]: I think solution should be in fp16, we should replace it with fp16 inf (== -65505) as valNum ?

Comment (krzysz00) [reply mlir/lib/ExecutionEngine/conv-validation-wrappers.cpp]: I think the solution is entirely different: define the semantics of fp16 truncation to be saturating CPU-side.

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp [hunk 2]
```diff
@@ -2885,9 +2885,15 @@ static func::FuncOp createVerifierFunc(ModuleOp module, const KernelIF &kernel,
...
 
   if (testElemType.isa<FloatType>()) {
-    auto thr_RMS = getF32Val(RMSThreshold.getValue());
+    constexpr float defaultRMSThreshold(0.00003f);
+    constexpr float defaultRMSThresholdFP16(0.001f);
+    float RMSThresholdValue =
+        testElemType.isF16() ? defaultRMSThresholdFP16 : defaultRMSThreshold;
+    if (RMSThreshold)
+      RMSThresholdValue = RMSThreshold.getValue();
+    auto thr_RMS = getF32Val(RMSThresholdValue);
```
Comment (krzysz00) [inline mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: Can we not `auto` this?

File: mlir/lib/ExecutionEngine/conv-validation-wrappers.cpp (hunk 1 above)
Comment (krzysz00) [reply mlir/lib/ExecutionEngine/conv-validation-wrappers.cpp]: ... Wait no, I see yet another solution: when we generate FP16 verification kernels, explicitly set the rounding mode to `rtz` around the conversions

**Review by krzysz00 — COMMENTED** [review]
I've got Thoughts here - maybe we land this anyway to unbreak CI but still

File: mlir/tools/rocmlir-gen/rocmlir-gen.cpp (hunk 2 above)
Comment (giuseros) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: What do you mean `auto`?

Comment (krzysz00) [reply mlir/tools/rocmlir-gen/rocmlir-gen.cpp]: You're using `auto` in a context where it's easy and helps clarity to write the type out.

(The existing code used `Value` for one of these constants, and you even changed it to `auto`)

**Review by krzysz00 — APPROVED** [review]

**Pull request merged**